contrib/endpoints/src/api_manager/check_security_rules.cc

```diff
@@ -14,92 +14,335 @@
+      status = invalid;
+    } else if (std::string(result) != kTestSuccess) {
+      status = Status(Code::PERMISSION_DENIED,
+                      std::string("Unauthorized ") +
+                          context.request()->GetRequestHTTPMethod() +
+                          " access to resource " +
+                          context.request()->GetRequestPath(),
+                      Status::AUTH);
+    }
+  }
+
+  grpc_json_destroy(json);
+  return status;
+}
+
+Status AuthzChecker::BuildTestRequestBody(context::RequestContext &context,
+                                          std::string *result_string) {
+  proto::TestRulesetRequest request;
+  auto *test_case = request.add_test_cases();
+  auto httpMethod = context.request()->GetRequestHTTPMethod();
+
+  test_case->set_service_name(context.service_context()->service_name());
+  test_case->set_resource_path(context.request()->GetRequestPath());
+  test_case->set_operation(GetOperation(httpMethod));
+  test_case->set_expectation(proto::TestRulesetRequest::TestCase::ALLOW);
+
+  ::google::protobuf::Value auth;
+  ::google::protobuf::Value token;
+  ::google::protobuf::Value claims;
+
+  Status status = utils::JsonToProto(context.auth_claims(), &claims);
+  if (!status.ok()) {
+    env_->LogError(std::string("Error creating Protobuf from claims") +
+                   status.ToString());
+    return status;
+  }
+
+  SetProtoValue("token", claims, &token);
+  SetProtoValue("auth", token, &auth);
+
+  auto *variables = test_case->mutable_variables();
+  (*variables)["request"] = auth;
+
+  status =
+      utils::ProtoToJson(request, result_string, utils::JsonOptions::DEFAULT);
+  if (status.ok()) {
+    env_->LogDebug(std::string("PRotobuf to JSON string = ") + *result_string);
+  } else {
+    env_->LogError(std::string("Error creating TestRulesetRequest") +
+                   status.ToString());
+  }
+
+  return status;
 }
 
 void AuthzChecker::HttpFetch(
-    const std::string &url, const std::string &request_body,
+    const std::string &url, const std::string &method,
+    const std::string &request_body,
     std::function<void(Status, std::string &&)> continuation) {
+  env_->LogDebug(std::string("Issue HTTP Request to url :") + url +
+                 " method : " + method + " body: " + request_body);
+
   std::unique_ptr<HTTPRequest> request(new HTTPRequest([continuation](
       Status status, std::map<std::string, std::string> &&,
       std::string &&body) { continuation(status, std::move(body)); }));
+
   if (!request) {
     continuation(Status(Code::INTERNAL, "Out of memory"), "");
     return;
   }
 
-  request->set_method("POST")
-      .set_url(url)
-      .set_auth_token(GetAuthToken())
-      .set_header("Content-Type", "application/json")
-      .set_body(request_body);
+  request->set_method(method).set_url(url).set_auth_token(GetAuthToken());
+
+  if (method != "GET") {
+    request->set_header("Content-Type", "application/json")
```

Comment:
I have removed as many string constants from code as I could. Let me know if you think there are more that need to be changed.